<div style="
    display:flex;
    justify-content:center;
    align-items:center;
    width:72%;
    margin:40px auto;
    border-radius:14px;
    background: linear-gradient(45deg, #8a2be2, #ff3cac, #2bd2ff, #8a2be2);
    background-size: 300% 300%;
    animation: borderFlow 7s linear infinite;
    padding:2px;
    box-sizing:border-box;
">
  <div style="
      display:flex;
      align-items:center;
      justify-content:center;
      flex-wrap:wrap;
      background: linear-gradient(90deg, rgba(18,18,28,0.95), rgba(28,18,40,0.95));
      color: white;
      font-family: 'Segoe UI', Arial, sans-serif;
      font-size: 26px;
      font-weight: 650;
      letter-spacing: 1px;
      padding: 18px 34px;
      border-radius: 12px;
      text-transform: uppercase;
      box-shadow: inset 0 0 0 0 rgba(0,0,0,0), 0 3px 10px rgba(0,0,0,0.35);
      width: 100%;
      height: 100%;
      box-sizing:border-box;
  ">
    <img src="assets/img/lapage.png"
         alt="Logo Lapage"
         style="height:110px; width:auto; filter: drop-shadow(0 0 10px rgba(138,43,226,0.35)); margin-right:22px;">
    <div style="text-align:center;">
      Analyse des ventes & du comportement clients<br>
      <span style="font-size:18px; opacity:0.85; letter-spacing:0.5px; text-transform:none;">
        Data-driven insights for decision making
      </span>
    </div>
  </div>
</div>

<style>
@keyframes borderFlow {
  0% { background-position: 0% 50%; }
  50% { background-position: 100% 50%; }
  100% { background-position: 0% 50%; }
}
</style>

# 🎯 Objectif du projet — Lapage 📚

Ce notebook a pour objectif d’**analyser les ventes et le comportement des clients de l’entreprise Lapage**, une librairie ayant développé son activité de vente en ligne, afin d’en extraire des **insights clairs et exploitables pour le comité de direction (CODIR)**.

L’analyse vise à fournir une **vision synthétique et stratégique de la performance commerciale**, ainsi qu’une meilleure compréhension des profils et habitudes des clients.

---

## 🧩 Contexte

Initialement positionnée comme une **librairie physique**, l’entreprise **Lapage** a lancé son site de vente en ligne il y a deux ans afin de répondre à la demande croissante de ses clients et d’élargir son activité.

Dans ce contexte, la direction souhaite aujourd’hui :

- mieux **suivre l’évolution de l’activité e-commerce** ;
- comprendre les **dynamiques de vente** ;
- analyser le **comportement des clients en ligne**.

Pour mener cette étude, trois jeux de données sont mis à disposition :

- un fichier **clients** (genre, année de naissance) ;
- un fichier **produits** (prix et catégories) ;
- un fichier **transactions** (dates d’achat, clients et produits).

---

## 📈 Objectifs principaux

Le travail se structure autour de **deux axes complémentaires**.

### Axe 1 — Analyse de l’activité commerciale

- Étudier l’**évolution du chiffre d’affaires dans le temps** ;
- Mettre en évidence les tendances à l’aide de **moyennes mobiles** ;
- Analyser le **nombre de clients**, de **transactions** et de **produits vendus** ;
- Identifier les **produits les plus performants (tops)** et les **moins performants (flops)** ;
- Étudier la **répartition des ventes par catégorie**.

---

### Axe 2 — Analyse du comportement client

- Étudier la **répartition du chiffre d’affaires par client** ;
- Analyser la **dispersion des ventes** à l’aide d’une **courbe de Lorenz** ;
- Mettre en évidence des **corrélations entre les variables clients** :
  - âge ;
  - genre ;
  - fréquence d’achat ;
  - panier moyen ;
  - catégories de livres achetées.
- Appliquer et justifier l’utilisation de **tests statistiques adaptés** :
  - test du **χ² (khi-deux)** ;
  - **ANOVA** ou **test de Student** ;
  - **corrélation de Pearson ou de Spearman**.
- Identifier et traiter les **clients atypiques (BtoB)** afin de garantir la pertinence des analyses.

---

## 🧠 Finalité

Les résultats issus de cette étude permettront :

- d’**accompagner la prise de décision stratégique du CODIR** ;
- d’orienter les **choix marketing et commerciaux** (offres, ciblage, positionnement) ;
- de proposer une **lecture claire, pédagogique et exploitable des données**, adaptée à un public non technique.

---

> 💡 *Cette analyse a été menée dans une démarche de rigueur, de clarté et de reproductibilité, afin de fournir à Lapage une vision fiable et structurée de son activité e-commerce et du comportement de ses clients.*



<a id="top"></a>

<h2 style="color:RGB(170,90,255);">📘 Sommaire</h2>

<ul style="list-style:none; line-height:1.8; font-size:17px; padding-left:0;">

  <li>
    <a href="#etape1" style="color:RGB(170,90,255); text-decoration:none;">📂 Étape 1 — Importation et préparation des données</a>
    <ul style="list-style:none; margin-left:20px;">
      <li><a href="#etape1_1" style="color:RGB(170,90,255); text-decoration:none;">1.1 Importation des librairies</a></li>
      <li><a href="#etape1_2" style="color:RGB(170,90,255); text-decoration:none;">1.2 Chargement des fichiers CSV</a></li>
    </ul>
  </li>

  <li>
  <a href="#etape2" style="color:RGB(120,210,255); text-decoration:none;">🔎 Étape 2 — Analyse exploratoire des données (EDA)</a>
  <ul style="list-style:none; margin-left:20px;">
    <li><a href="#etape2_1" style="color:RGB(120,210,255); text-decoration:none;">2.1 Analyse du fichier clients</a></li>
    <li><a href="#etape2_2" style="color:RGB(120,210,255); text-decoration:none;">2.2 Analyse du fichier produits</a></li>
    <li><a href="#etape2_3" style="color:RGB(120,210,255); text-decoration:none;">2.3 Analyse du fichier transactions</a></li>
    <li><a href="#etape2_3_1" style="margin-left:2rem; color:RGB(120,210,255); text-decoration:none;">2.3.1 Analyse des doublons dans le fichier transactions</a></li>
    <li><a href="#etape2_3_2" style="margin-left:2rem; color:RGB(120,210,255); text-decoration:none;">2.3.2 Traitements des doublons dans le fichier transactions</a></li>
  </ul>
</li>


  <li>
    <a href="#etape3" style="color:RGB(255,170,80); text-decoration:none;">🔗 Étape 3 — Construction du jeu de données d’analyse</a>
    <ul style="list-style:none; margin-left:20px;">
      <li><a href="#etape3_1" style="color:RGB(255,170,80); text-decoration:none;">3.1 Jointure transactions + produits</a></li>
      <li><a href="#etape3_2" style="color:RGB(255,170,80); text-decoration:none;">3.2 Jointure avec les données clients</a></li>
      <li><a href="#etape3_3" style="color:RGB(255,170,80); text-decoration:none;">3.3 Validation du dataset final</a></li>
    </ul>
  </li>

  <li>
    <a href="#etape4" style="color:RGB(120,255,170); text-decoration:none;">📈 Étape 4 — Analyse temporelle des ventes</a>
    <ul style="list-style:none; margin-left:20px;">
      <li><a href="#etape4_1" style="color:RGB(120,255,170); text-decoration:none;">4.1 Évolution du chiffre d’affaires</a></li>
      <li><a href="#etape4_2" style="color:RGB(120,255,170); text-decoration:none;">4.2 Moyennes mobiles</a></li>
      <li><a href="#etape4_3" style="color:RGB(120,255,170); text-decoration:none;">4.3 Clients, transactions, produits par période</a></li>
    </ul>
  </li>

  <li>
    <a href="#etape5" style="color:RGB(255,120,190); text-decoration:none;">🏆 Étape 5 — Analyse de la performance des produits</a>
    <ul style="list-style:none; margin-left:20px;">
      <li><a href="#etape5_1" style="color:RGB(255,120,190); text-decoration:none;">5.1 Tops produits</a></li>
      <li><a href="#etape5_2" style="color:RGB(255,120,190); text-decoration:none;">5.2 Flops produits</a></li>
      <li><a href="#etape5_3" style="color:RGB(255,120,190); text-decoration:none;">5.3 Répartition du CA par catégorie</a></li>
    </ul>
  </li>

  <li>
    <a href="#etape6" style="color:RGB(255,210,90); text-decoration:none;">🧑‍🤝‍🧑 Étape 6 — Analyse des clients</a>
    <ul style="list-style:none; margin-left:20px;">
      <li><a href="#etape6_1" style="color:RGB(255,210,90); text-decoration:none;">6.1 Profils clients & données démographiques</a></li>
      <li><a href="#etape6_2" style="color:RGB(255,210,90); text-decoration:none;">6.2 Concentration du CA & courbe de Lorenz</a></li>
    </ul>
  </li>

  <li>
    <a href="#etape7" style="color:RGB(180,255,120); text-decoration:none;">🧪 Étape 7 — Corrélations & tests statistiques</a>
    <ul style="list-style:none; margin-left:20px;">
      <li><a href="#etape7_1" style="color:RGB(180,255,120); text-decoration:none;">7.1 Corrélations exploratoires</a></li>
      <li><a href="#etape7_2" style="color:RGB(180,255,120); text-decoration:none;">7.2 Tests : χ², ANOVA (ou t-test), Pearson/Spearman</a></li>
      <li><a href="#etape7_3" style="color:RGB(180,255,120); text-decoration:none;">7.3 Interprétation des résultats</a></li>
    </ul>
  </li>

  <li>
    <a href="#etape8" style="color:RGB(255,120,80); text-decoration:none;">⚠️ Étape 8 — Clients atypiques (BtoB)</a>
    <ul style="list-style:none; margin-left:20px;">
      <li><a href="#etape8_1" style="color:RGB(255,120,80); text-decoration:none;">8.1 Identification des clients atypiques</a></li>
      <li><a href="#etape8_2" style="color:RGB(255,120,80); text-decoration:none;">8.2 Justification de l’exclusion pour l’analyse statistique</a></li>
      <li><a href="#etape8_3" style="color:RGB(255,120,80); text-decoration:none;">8.3 Nouveau dataset “sans BtoB”</a></li>
    </ul>
  </li>

  <li>
    <a href="#conclusion" style="color:RGB(170,170,170); text-decoration:none;">✅ Conclusion & recommandations métier</a>
  </li>

</ul>

<a id="etape2"></a>
<div style="
  border: 1px solid #6D28D9;
  border-radius: 8px;
  background-color: rgba(79,70,229,0.10);
  margin: auto;
">
  <h2 style="
    margin: auto;
    padding: 15px 20px;
    background-color: #4F46E5;
    color: #fff;
    display: flex;
    align-items: center;
    border-radius: 8px;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="30" style="vertical-align:middle;
        filter: invert(100%) brightness(110%);
      ">
    </a>
    Étape 1 — Importation et préparation des données
  </h2>
</div>

<a id="etape1_1"></a>
<div style="
  border: 1px solid #6D28D9;
  border-radius: 8px;
  background-color: rgba(109,40,217,0.06);
">
  <h3 style="
    margin: auto;
    padding: 12px 18px;
    color: #4F46E5;
    display: flex;
    align-items: center;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16" style="vertical-align:middle;
        filter: invert(28%) sepia(86%) saturate(2421%) hue-rotate(252deg) brightness(94%) contrast(96%);
      ">
    </a>
    1.1 - Importation des librairies
  </h3>
</div>

In [224]:
#Importation de la librairie Pandas
from IPython.display import display, HTML
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sns
import warnings
warnings.filterwarnings("ignore", category=UserWarning, module="openpyxl")

print("✅ Tout est installé et prêt 🎉")

✅ Tout est installé et prêt 🎉


In [225]:
# Affichage de toutes les colonnes du dataframe
pd.set_option('display.max_columns', None)

<div style="background-color:#1e1e1e; color:#dcdcdc; padding:10px 15px; border-radius:8px; font-family:Consolas, monospace; border:1px solid #444; width:fit-content; margin:auto;">

<b>Réponse :</b><br>
<span style="color:#9CDCFE;">pd</span>.<span style="color:#DCDCAA;">set_option</span>(
<span style="color:#CE9178;">'display.max_columns'</span>, 
<span style="color:#569CD6;">None</span>)
<div style="color:#32CD32; font-weight:bold; font-family:Segoe UI, sans-serif; font-size:15px; text-align:center; margin-top:10px;">
✅ Toutes les colonnes du DataFrame seront désormais affichées.
</div>
</div>

<a id="etape1_2"></a>
<div style="
  border: 1px solid #6D28D9;
  border-radius: 8px;
  background-color: rgba(109,40,217,0.06);
">
  <h3 style="
    margin: auto;
    padding: 12px 18px;
    color: #4F46E5;
    display: flex;
    align-items: center;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16" style="vertical-align:middle;
        filter: invert(28%) sepia(86%) saturate(2421%) hue-rotate(252deg) brightness(94%) contrast(96%);
      ">
    </a>
    1.2 - Chargements des fichiers
  </h3>
</div>

In [226]:
#Importation du fichier customers.csv
customers = pd.read_csv(
    "Data/customers.csv",
    sep=";",
    dtype={"client_id": "string", "sex": "string", "birth": "Int64"},
    low_memory=False
)

#Importation du fichier products.csv
products = pd.read_csv(
    "Data/products.csv",
    sep=";",
    dtype={"id_prod": "string", "price": "float64", "categ": "Int64"},
    low_memory=False
)

#Importation du fichier transactions.csv
transactions = pd.read_csv(
    "Data/transactions.csv",
    sep=";",
    dtype={"id_prod": "string", "session_id": "string", "client_id": "string"},
    parse_dates=["date"],
    low_memory=False
)

<div style="
  background:#4CAF50;
  color:black;
  padding:15px 40px;
  border-radius:10px;
  font-family:Segoe UI;
  font-size:16px;
  text-align:center;
  width:fit-content;
  margin:30px auto;
">
<b>Résultat :</b> 🔥 Importation terminée :<br>
 les fichiers <b>customers.csv</b>, <b>products.csv</b> et <b>transactions.csv</b> sont chargés avec succès !
</div>

<a id="etape2"></a>
<div style="
  border: 1px solid #6D28D9;
  border-radius: 8px;
  background-color: rgba(79,70,229,0.10);
  margin: auto;
">
  <h2 style="
    margin: auto;
    padding: 15px 20px;
    background-color: #4F46E5;
    color: #fff;
    display: flex;
    align-items: center;
    border-radius: 8px;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="30" style="vertical-align:middle;
        filter: invert(100%) brightness(110%);
      ">
    </a>
    Etape 2 - Analyse exploratoire des fichiers
  </h2>
</div>

<a id="etape2_1"></a>
<div style="
  border: 1px solid #6D28D9;
  border-radius: 8px;
  background-color: rgba(109,40,217,0.06);
">
  <h3 style="
    margin: auto;
    padding: 12px 18px;
    color: #4F46E5;
    display: flex;
    align-items: center;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16" style="vertical-align:middle;
        filter: invert(28%) sepia(86%) saturate(2421%) hue-rotate(252deg) brightness(94%) contrast(96%);
      ">
    </a>
    2.1 - Analyse exploratoire du fichier customer.csv
  </h3>
</div>

In [227]:
#Afficher les dimensions du dataset
rows, cols = customers.shape
display(HTML(f"""
<p style='font-family: Segoe UI, Arial; font-size:18px; color:#e0e0e0;'>
Le tableau comporte <span style='color:#ffcc00; font-weight:bold;'>{rows:,}</span> observation(s) ou article(s)<br>
Le tableau comporte <span style='color:#ffcc00; font-weight:bold;'>{cols}</span> colonne(s)
</p>
"""))

In [228]:
# Focus sur un client (approche qualitative)
customers.sample(1)

,client_id,sex,birth
393,c_3935,f,1975


In [229]:
# Types de variables & valeurs manquantes
customers.info()
customers.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8621 entries, 0 to 8620
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   client_id  8621 non-null   string
 1   sex        8621 non-null   string
 2   birth      8621 non-null   Int64 
dtypes: Int64(1), string(2)
memory usage: 210.6 KB


client_id    0
sex          0
birth        0
dtype: int64

In [230]:
from IPython.display import display, HTML

# ✅ Vérification d'unicité & doublons sur l'identifiant client
is_unique = customers["client_id"].is_unique
dup_count = customers.duplicated(subset=["client_id"]).sum()

display(HTML(f"""
<p style="font-family: Segoe UI, Arial; font-size:18px; color:#e0e0e0;">
• Unicité de <b style="color:#ffcc00;">client_id</b> : 
<b style="color:{'#34d399' if is_unique else '#f87171'};">{is_unique}</b><br>
• Nombre de doublons sur <b style="color:#ffcc00;">client_id</b> : 
<b style="color:{'#34d399' if dup_count==0 else '#f87171'};">{dup_count}</b>
</p>
"""))

In [231]:
# Analyse des variables catégorielles (sex)
sex_counts = customers['sex'].value_counts()
sex_percent = customers['sex'].value_counts(normalize=True) * 100

df_sex = pd.DataFrame({
    "Effectif": sex_counts,
    "Pourcentage (%)": sex_percent.round(2)
})

df_sex

,Effectif,Pourcentage (%)
sex,,
f,4490,52.08
m,4131,47.92


In [232]:
# Analyse descriptive de l’âge
ANNEE_ANALYSE = 2026

customers['age'] = ANNEE_ANALYSE - customers['birth']
customers[['birth', 'age']].head()

,birth,age
0,1967,59
1,1975,51
2,1984,42
3,1962,64
4,1943,83


In [233]:
# Vérifier les valeurs possibles de sex (qualité)
customers["sex"].value_counts(dropna=False)

sex
f    4490
m    4131
Name: count, dtype: Int64

In [234]:
f_pct = df_sex.loc["f", "Pourcentage (%)"]
m_pct = df_sex.loc["m", "Pourcentage (%)"]

display(HTML(f"""
<p style="font-family: Segoe UI, Arial; font-size:18px; color:#e0e0e0;">
Répartition : <b style="color:#ffcc00;">Femmes</b> = <b>{f_pct}%</b> • 
<b style="color:#ffcc00;">Hommes</b> = <b>{m_pct}%</b>
</p>
"""))

In [235]:
import plotly.express as px

fig = px.histogram(customers, x="age", nbins=20, title="Distribution de l'âge des clients")
fig.update_layout(template="plotly_dark")
fig.show()

<div style="
  font-family: Segoe UI, Arial;
  font-size: 16.5px;
  color: #eaeaea;
  background: linear-gradient(180deg, rgba(30,30,30,0.92), rgba(18,18,18,0.92));
  padding: 18px 20px;
  border-radius: 12px;
  border: 1px solid rgba(120,86,255,0.35);
  width: min(900px, 95%);
  margin: 22px auto;
  text-align: left;
  box-shadow: 0 0 18px rgba(0,0,0,0.45);
">

  <div style="display:flex; align-items:center; gap:10px; margin-bottom:8px;">
    <span style="font-size:18px;">👥</span>
    <b style="font-size:18px;">
      Synthèse — Analyse exploratoire du fichier 
      <span style="color:#ffcc00;">customers.csv</span>
    </b>
  </div>

  <hr style="border:none; border-top:1px solid rgba(255,255,255,0.08); margin:10px 0 14px 0;" />

  <p style="margin:8px 0;">
    📌 <b>Dimensions :</b> 
    <span style="color:#ffcc00; font-weight:700;">8 621</span> lignes • 
    <span style="color:#ffcc00; font-weight:700;">3</span> colonnes
    <span style="color:#9aa0a6;">(client_id, sex, birth)</span>
  </p>

  <p style="margin:8px 0;">
    🧬 <b>Types :</b> <span style="color:#c9d1d9;">client_id</span> (string), 
    <span style="color:#c9d1d9;">sex</span> (string), 
    <span style="color:#c9d1d9;">birth</span> (int)
  </p>

  <p style="margin:8px 0;">
    ✅ <b>Valeurs manquantes :</b> <span style="color:#34d399; font-weight:700;">0</span> 
    <span style="color:#9aa0a6;">(aucune valeur null détectée)</span>
  </p>

  <p style="margin:8px 0;">
    🔎 <b>Qualité / Identifiants :</b><br>
    • Unicité de <span style="color:#ffcc00; font-weight:700;">client_id</span> : 
      <span style="color:#34d399; font-weight:700;">True</span><br>
    • Doublons sur <span style="color:#ffcc00; font-weight:700;">client_id</span> :
      <span style="color:#34d399; font-weight:700;">0</span>
  </p>

  <p style="margin:8px 0;">
    🚻 <b>Répartition de la variable <span style="color:#ffcc00;">sex</span> :</b><br>
    • Femmes : <span style="color:#ffcc00; font-weight:700;">52.08%</span><br>
    • Hommes : <span style="color:#ffcc00; font-weight:700;">47.92%</span><br>
    <span style="color:#9aa0a6;">(modalités observées : f, m)</span>
  </p>

  <p style="margin:8px 0;">
    🎂 <b>Création de la variable <span style="color:#ffcc00;">age</span> :</b> 
    <span style="color:#9aa0a6;">age = 2026 − birth</span><br>
    • Âge moyen : <span style="color:#ffcc00; font-weight:700;">47.72</span> ans<br>
    • Médiane : <span style="color:#ffcc00; font-weight:700;">47</span> ans<br>
    • Min : <span style="color:#ffcc00; font-weight:700;">22</span> ans • 
      Max : <span style="color:#ffcc00; font-weight:700;">97</span> ans
  </p>

  <div style="
    margin-top: 14px;
    padding: 10px 8.4px;
    border-radius: 10px;
    background-color: rgba(79,70,229,0.10);
    border: 1px solid rgba(79,70,229,0.25);
  ">
    ✅ <b>Conclusion :</b> le fichier <span style="color:#ffcc00;"><b>customers.csv</b></span> est 
    <b>cohérent</b> (pas de valeurs manquantes, pas de doublons sur l’identifiant) et peut être
    utilisé pour les jointures ultérieures avec la table <b>transactions</b>.
  </div>

</div>

<a id="etape2_2"></a>
<div style="
  border: 1px solid #6D28D9;
  border-radius: 8px;
  background-color: rgba(109,40,217,0.06);
">
  <h3 style="
    margin: auto;
    padding: 12px 18px;
    color: #4F46E5;
    display: flex;
    align-items: center;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16" style="vertical-align:middle;
        filter: invert(28%) sepia(86%) saturate(2421%) hue-rotate(252deg) brightness(94%) contrast(96%);
      ">
    </a>
    2.2 - Analyse exploratoire du fichier Products.csv
  </h3>
</div>

In [236]:
# Dimensions & structure
rows, cols = products.shape
products.info()
products.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3286 entries, 0 to 3285
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id_prod  3286 non-null   string 
 1   price    3286 non-null   float64
 2   categ    3286 non-null   Int64  
dtypes: Int64(1), float64(1), string(1)
memory usage: 80.4 KB


id_prod    0
price      0
categ      0
dtype: int64

<div style="
  font-family:Segoe UI, Arial;
  font-size:17px;
  color:#e0e0e0;
  background-color:#1e1e1e;
  padding:20px;
  border-radius:10px;
  border:1px solid #444;
  width:fit-content;
  margin:20px auto;
">
<p>✅ <b>Valeurs manquantes</b></p>
<p>Aucune valeur manquante n’a été détectée dans le fichier <b>Products</b>.<br>
Le dataset est <b>complet</b> et directement exploitable pour les analyses suivantes.</p>
</div>

In [237]:
# Unicité & doublons
is_unique_prod = products["id_prod"].is_unique
dup_prod = products.duplicated(subset=["id_prod"]).sum()

is_unique_prod, dup_prod = products["id_prod"].is_unique, products.duplicated(subset=["id_prod"]).sum()

print(f"Unicité de id_prod : {is_unique_prod}")
print(f"Nombre de doublons sur id_prod : {dup_prod}")

if dup_prod == 0:
    print("✅ Aucun doublon détecté.")
else:
    print("⚠️ Doublons détectés : à investiguer avant les jointures.")

Unicité de id_prod : True
Nombre de doublons sur id_prod : 0
✅ Aucun doublon détecté.


<div style="
  font-family:Segoe UI, Arial;
  font-size:17px;
  color:#e0e0e0;
  background-color:#1e1e1e;
  padding:18px 22px;
  border-radius:12px;
  border:1px solid #3a3a3a;
  width:fit-content;
  margin:18px auto;
  box-shadow:0 0 15px #00000055;
  text-align:left;
">
  <p style="margin:0 0 10px 0;">
    📦 <b>Vérification de l’unicité des produits</b>
  </p>

  <ul style="margin:0; padding-left:18px; line-height:1.55;">
    <li>
      Unicité de <b style="color:#ffcc00;">id_prod</b> :
      <b style="color:#34d399;">True</b>
    </li>
    <li>
      Nombre de doublons sur <b style="color:#ffcc00;">id_prod</b> :
      <b style="color:#34d399;">0</b>
    </li>
  </ul>

  <div style="margin-top:10px; font-weight:600; color:#34d399;">
    ✅ Aucun doublon détecté.
  </div>
</div>

In [238]:
# Analyse des prix
products['price'].describe()

count    3286.000000
mean       21.863597
std        29.849786
min         0.620000
25%         6.990000
50%        13.075000
75%        22.990000
max       300.000000
Name: price, dtype: float64

In [239]:
# Analyse des prix (tableau)
categ_counts = products["categ"].value_counts()
categ_pct = (products["categ"].value_counts(normalize=True) * 100).round(2)

df_categ = pd.DataFrame({
    "Effectif": categ_counts,
    "Pourcentage (%)": categ_pct
})
df_categ

,Effectif,Pourcentage (%)
categ,,
0,2308,70.24
1,739,22.49
2,239,7.27


In [240]:
# Graphique Analyse des prix
fig = px.box(products, y="price", title="Distribution des prix des produits")
fig.update_layout(template="plotly_dark")
fig.show()

<div style="
  font-family:Segoe UI, Arial;
  font-size:16px;
  color:#e0e0e0;
  background:#1e1e1e;
  padding:16px 20px;
  border-radius:12px;
  border:1px solid #3a3a3a;
  max-width:860px;
  margin:14px auto 22px auto;
  box-shadow:0 0 15px #00000055;
">
  <p style="margin:0 0 10px 0;">
    📦 <b>Lecture du boxplot — Distribution des prix</b>
  </p>

  <ul style="margin:0; padding-left:18px; line-height:1.55;">
    <li>La majorité des produits se situe dans une zone de prix relativement basse (boîte proche de 0–30€).</li>
    <li>La distribution est <b>asymétrique à droite</b> : quelques produits ont des prix beaucoup plus élevés que la moyenne.</li>
    <li>Les points au-dessus de la moustache représentent des <b>valeurs atypiques</b> (produits “premium” ou cas spécifiques).</li>
  </ul>

  <p style="margin:10px 0 0 0; color:#bdbdbd;">
    ✅ Ces valeurs élevées ne sont pas supprimées à ce stade : elles peuvent être cohérentes métier (gamme premium).
  </p>
</div>

In [241]:
# Analyse des catégories
products['categ'].value_counts()
products['categ'].value_counts(normalize=True) * 100

categ
0    70.237371
1    22.489349
2     7.273281
Name: proportion, dtype: Float64

<div style="
  font-family:Segoe UI, Arial;
  font-size:16px;
  color:#e0e0e0;
  background:#1e1e1e;
  padding:16px 20px;
  border-radius:12px;
  border:1px solid #3a3a3a;
  max-width:700px;
  margin:16px auto;
">
<p style="margin-top:0;">📊 <b>Lecture analytique des catégories</b></p>

<ul>
  <li><b>Catégorie 0</b> : majorité des produits (~70%), prix faibles et homogènes → entrée de gamme.</li>
  <li><b>Catégorie 1</b> : gamme intermédiaire, prix moyens plus élevés.</li>
  <li><b>Catégorie 2</b> : produits premium, prix élevés et forte dispersion.</li>
</ul>

<p style="margin-bottom:0;">
👉 Les différences de prix observées sont cohérentes avec une segmentation par gamme.
</p>
</div>

In [242]:
# Vérification cohérence catégorie ↔ prix
products.groupby('categ')['price'].describe()

,count,mean,std,min,25%,50%,75%,max
categ,,,,,,,,
0,2308.0,11.732795,7.565755,0.62,5.5875,10.32,16.655,40.99
1,739.0,25.531421,15.425162,2.00,13.3900,22.99,33.990,80.99
2,239.0,108.354686,49.561431,30.99,71.0650,101.99,136.530,300.00


<div style="
  font-family:Segoe UI, Arial;
  font-size:16px;
  color:#e0e0e0;
  background-color:#1e1e1e;
  padding:18px 22px;
  border-radius:12px;
  border:1px solid #3a3a3a;
  width:fit-content;
  margin:18px auto;
  box-shadow:0 0 15px #00000055;
  line-height:1.5;
">
  <p style="margin:0 0 10px 0;">
    🧾 <b>Lecture du tableau <span style="color:#ffcc00;">describe()</span> par catégorie</b>
  </p>

  <ul style="margin:0; padding-left:18px;">
    <li><b style="color:#ffcc00;">count</b> : nombre de produits dans la catégorie.</li>
    <li><b style="color:#ffcc00;">mean</b> : prix <b>moyen</b> des produits de la catégorie.</li>
    <li><b style="color:#ffcc00;">std</b> : <b>écart-type</b> (dispersion des prix). 
      <span style="color:#9ca3af;">Plus il est élevé, plus les prix sont variés.</span>
    </li>
    <li><b style="color:#ffcc00;">min</b> / <b style="color:#ffcc00;">max</b> : prix minimum / maximum observés.</li>
    <li><b style="color:#ffcc00;">25%</b> / <b style="color:#ffcc00;">50%</b> / <b style="color:#ffcc00;">75%</b> : quartiles
      (valeurs seuils : 25%, médiane, 75% des prix).</li>
  </ul>

  <p style="margin:10px 0 0 0; color:#9ca3af;">
    💡 Exemple : si <b>50%</b> = 10.32, alors la moitié des produits de la catégorie ont un prix ≤ 10.32.
  </p>
</div>

In [243]:
# Focus produit (qualitatif)
products.sample(1)

,id_prod,price,categ
1058,1_153,22.99,1


<div style="
  font-family: Segoe UI, Arial;
  font-size: 16.5px;
  color: #eaeaea;
  background: linear-gradient(180deg, rgba(30,30,30,0.92), rgba(18,18,18,0.92));
  padding: 18px 20px;
  border-radius: 12px;
  border: 1px solid rgba(120,86,255,0.35);
  width: min(900px, 95%);
  margin: 22px auto;
  text-align: left;
  box-shadow: 0 0 18px rgba(0,0,0,0.45);
">

  <div style="display:flex; align-items:center; gap:10px; margin-bottom:8px;">
    <span style="font-size:18px;">📦</span>
    <b style="font-size:18px;">
      Synthèse — Analyse exploratoire du fichier 
      <span style="color:#ffcc00;">Products.csv</span>
    </b>
  </div>

  <hr style="border:none; border-top:1px solid rgba(255,255,255,0.08); margin:10px 0 14px 0;" />

  <p style="margin:8px 0;">
    📌 <b>Dimensions :</b> 
    <span style="color:#ffcc00; font-weight:700;">3 286</span> lignes • 
    <span style="color:#ffcc00; font-weight:700;">3</span> colonnes
    <span style="color:#9aa0a6;">(id_prod, price, categ)</span>
  </p>

  <p style="margin:8px 0;">
    🧬 <b>Types :</b> <span style="color:#c9d1d9;">id_prod</span> (string), 
    <span style="color:#c9d1d9;">price</span> (float), 
    <span style="color:#c9d1d9;">categ</span> (int)
  </p>

  <p style="margin:8px 0;">
    ✅ <b>Valeurs manquantes :</b> <span style="color:#34d399; font-weight:700;">0</span> 
    <span style="color:#9aa0a6;">(aucune valeur null détectée)</span>
  </p>

  <p style="margin:8px 0;">
    🔎 <b>Qualité / Identifiants :</b><br>
    • Unicité de <span style="color:#ffcc00; font-weight:700;">id_prod</span> : 
      <span style="color:#34d399; font-weight:700;">True</span><br>
    • Doublons sur <span style="color:#ffcc00; font-weight:700;">id_prod</span> :
      <span style="color:#34d399; font-weight:700;">0</span>
  </p>

  <p style="margin:8px 0;">
    💰 <b>Analyse de la variable <span style="color:#ffcc00;">price</span> :</b><br>
    • Prix minimum : <span style="color:#ffcc00; font-weight:700;">0.62</span><br>
    • Médiane (50%) : <span style="color:#ffcc00; font-weight:700;">13.08</span><br>
    • 25% des produits ≤ <span style="color:#ffcc00; font-weight:700;">6.99</span> • 
      75% des produits ≤ <span style="color:#ffcc00; font-weight:700;">22.99</span><br>
    • Prix maximum : <span style="color:#ffcc00; font-weight:700;">300.00</span><br>
    <span style="color:#9aa0a6;">(distribution asymétrique : présence de produits “premium” / outliers)</span>
  </p>

  <p style="margin:8px 0;">
    🏷️ <b>Répartition de la variable <span style="color:#ffcc00;">categ</span> :</b><br>
    • Catégorie 0 : <span style="color:#ffcc00; font-weight:700;">70.24%</span> 
      <span style="color:#9aa0a6;">(2 308 produits)</span><br>
    • Catégorie 1 : <span style="color:#ffcc00; font-weight:700;">22.49%</span> 
      <span style="color:#9aa0a6;">(739 produits)</span><br>
    • Catégorie 2 : <span style="color:#ffcc00; font-weight:700;">7.27%</span> 
      <span style="color:#9aa0a6;">(239 produits)</span>
  </p>

  <div style="
    margin-top: 14px;
    padding: 10px 8.4px;
    border-radius: 10px;
    background-color: rgba(79,70,229,0.10);
    border: 1px solid rgba(79,70,229,0.25);
  ">
    ✅ <b>Conclusion :</b> le fichier <span style="color:#ffcc00;"><b>Products.csv</b></span> est 
    <b>cohérent</b> (pas de valeurs manquantes, pas de doublons sur l’identifiant) et peut être
    utilisé pour les jointures ultérieures avec la table <b>transactions</b>.
  </div>

</div>

<a id="etape2_3"></a>
<div style="
  border: 1px solid #6D28D9;
  border-radius: 8px;
  background-color: rgba(109,40,217,0.06);
">
  <h3 style="
    margin: auto;
    padding: 12px 18px;
    color: #4F46E5;
    display: flex;
    align-items: center;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16" style="vertical-align:middle;
        filter: invert(28%) sepia(86%) saturate(2421%) hue-rotate(252deg) brightness(94%) contrast(96%);
      ">
    </a>
    2.3 - Analyse exploratoire du fichier Transactions.csv
  </h3>
</div>

In [244]:
# Structure & valeurs manquantes
# Dimensions & structure
rows, cols = transactions.shape
transactions.info()
transactions.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 4 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   id_prod     687534 non-null  string        
 1   date        687534 non-null  datetime64[ns]
 2   session_id  687534 non-null  string        
 3   client_id   687534 non-null  string        
dtypes: datetime64[ns](1), string(3)
memory usage: 32.0 MB


id_prod       361041
date          361041
session_id    361041
client_id     361041
dtype: int64

In [245]:
# Unicité globale des lignes
transactions.duplicated().sum()

np.int64(361040)

In [246]:
# Clé logique de transaction
dup_tx = transactions.duplicated(subset=["session_id", "id_prod"]).sum()
print(f"Doublons sur la clé (session_id, id_prod) : {dup_tx}")

Doublons sur la clé (session_id, id_prod) : 361895


In [247]:
# Vérification des identifiants isolés

# Unicité des sessions
transactions["session_id"].is_unique

# Unicité des clients (normalement False)
transactions["client_id"].is_unique

# Unicité des produits (normalement False)
transactions["id_prod"].is_unique

print(f"session_id unique ?  {transactions['session_id'].is_unique}")
print(f"client_id unique ?   {transactions['client_id'].is_unique}")
print(f"id_prod unique ?     {transactions['id_prod'].is_unique}")

session_id unique ?  False
client_id unique ?   False
id_prod unique ?     False


In [248]:
# Vérification de cohérence avec les autres tables (clé étrangère)
missing_prod = transactions[~transactions["id_prod"].isin(products["id_prod"])]
missing_prod.shape[0]

print(f"Transactions avec id_prod absent de Products : {missing_prod.shape[0]}")

Transactions avec id_prod absent de Products : 361041


In [249]:
# Clients existants
missing_clients = transactions[~transactions["client_id"].isin(customers["client_id"])]
missing_clients.shape[0]

361041

In [250]:
# Nombre de transactions par client
transactions["client_id"].value_counts().describe()

# Nombre de produits par session
transactions["session_id"].value_counts().describe()

count    345505.0
mean     1.989939
std      1.276025
min           1.0
25%           1.0
50%           2.0
75%           3.0
max          14.0
Name: count, dtype: Float64

In [251]:
# Nombre d'occurrences par client / session / produit
tx_repeat = (
    transactions
    .groupby(["client_id", "session_id", "id_prod"])
    .size()
    .reset_index(name="nb_occurrences")
)

tx_repeat.describe()

,nb_occurrences
count,686679.000000
mean,1.001245
std,0.035347
min,1.000000
25%,1.000000
50%,1.000000
75%,1.000000
max,3.000000


In [252]:
# Identification des cas où un même produit est acheté plusieurs fois
# par un même client au sein d'une même session.
# Objectif : distinguer les doublons techniques des achats multiples légitimes.

multi_buy = tx_repeat[tx_repeat["nb_occurrences"] > 1]
multi_buy.shape[0]

853

In [253]:
# Exploration des cas les plus fréquents d’achats multiples
# afin de vérifier s’il s’agit de comportements plausibles.
# Extraction des cas où nb_occurrences > 1 (achats répétés dans une même session)

multi_buy.sort_values("nb_occurrences", ascending=False).head(10)

,client_id,session_id,id_prod,nb_occurrences
480648,c_6385,s_43170,0_1440,3
480601,c_6385,s_217835,0_1440,3
671447,c_857,s_192744,0_1023,2
16460,c_120,s_134184,1_459,2
676451,c_8618,s_335147,2_37,2
676196,c_8616,s_155330,0_1578,2
1987,c_1018,s_69540,1_366,2
7036,c_1081,s_18698,1_262,2
7039,c_1081,s_192060,1_262,2
7381,c_1087,s_149473,0_1340,2


<div style="
  font-family: Segoe UI, Arial;
  font-size: 16px;
  color: #eaeaea;
  background: linear-gradient(180deg, rgba(30,30,30,0.92), rgba(18,18,18,0.92));
  padding: 18px 20px;
  border-radius: 12px;
  border: 1px solid rgba(120,86,255,0.35);
  width: min(900px, 95%);
  margin: 22px auto;
  box-shadow: 0 0 18px rgba(0,0,0,0.45);
">

  <div style="display:flex; align-items:center; gap:10px; margin-bottom:10px;">
    <span style="font-size:18px;">🧾</span>
    <b style="font-size:18px;">Analyse des répétitions dans les transactions</b>
  </div>

  <p style="margin:8px 0;">
    L’analyse porte sur le nombre d’occurrences d’un même produit acheté par un même client au sein d’une même session.
  </p>

  <ul style="margin:10px 0 10px 18px;">
    <li>La majorité des transactions apparaissent <b>une seule fois</b>.</li>
    <li>Un nombre très limité de cas présente plusieurs occurrences (au maximum 3).</li>
    <li>Ces répétitions correspondent à des <b>achats multiples légitimes</b>.</li>
  </ul>

  <div style="
    margin-top: 12px;
    padding: 8.8px;
    border-radius: 10px;
    background-color: rgba(34,197,94,0.12);
    border: 1px solid rgba(34,197,94,0.35);
  ">
    ✅ <b>Conclusion :</b> les répétitions observées ne constituent pas des erreurs de données.  
    Aucune suppression n’est effectuée à ce stade afin de préserver l’information métier.
  </div>

</div>

<a id="etape2_3_1"></a>
<div style="
  border: 1px solid #6D28D9;
  border-radius: 8px;
  background-color: rgba(109,40,217,0.06);
">
  <h3 style="
    margin: auto;
    padding: 12px 18px;
    color: #4F46E5;
    display: flex;
    align-items: center;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16" style="vertical-align:middle;
        filter: invert(28%) sepia(86%) saturate(2421%) hue-rotate(252deg) brightness(94%) contrast(96%);
      ">
    </a>
    2.3.1 - Analyse des doublons dans le fichier Transactions.csv
  </h3>
</div>

In [254]:
# Lignes avec id_prod présent
missing_prod = transactions.loc[transactions["id_prod"].notna() & ~transactions["id_prod"].isin(products["id_prod"])]
print("Transactions avec id_prod absent de Products (hors NaN) :", missing_prod.shape[0])

# Lignes avec client_id présent
missing_clients = transactions.loc[transactions["client_id"].notna() & ~transactions["client_id"].isin(customers["client_id"])]
print("Transactions avec client_id absent de Customers (hors NaN) :", missing_clients.shape[0])

Transactions avec id_prod absent de Products (hors NaN) : 0
Transactions avec client_id absent de Customers (hors NaN) : 0


In [255]:
# 1) Doublons exacts sur toute la ligne
dup_rows = transactions.duplicated().sum()
print("Doublons exacts (ligne complète) :", dup_rows)

# 2) Doublons sur (session_id, id_prod) MAIS lignes pas forcément identiques
dup_key = transactions.duplicated(subset=["session_id", "id_prod"]).sum()
print("Doublons sur la clé (session_id, id_prod) :", dup_key)

# Voir des exemples de répétitions
examples = transactions[transactions.duplicated(subset=["session_id","id_prod"], keep=False)] \
    .sort_values(["session_id","id_prod"]).head(20)
examples

Doublons exacts (ligne complète) : 361040
Doublons sur la clé (session_id, id_prod) : 361895


,id_prod,date,session_id,client_id
201065,0_1403,2021-10-05 06:41:12.627312,s_100238,c_7849
201067,0_1403,2021-10-05 06:44:23.963598,s_100238,c_7849
201369,0_1360,2021-10-05 13:33:28.245483,s_100388,c_7883
201386,0_1360,2021-10-05 13:56:17.131590,s_100388,c_7883
201731,0_1172,2021-10-05 21:49:22.798997,s_100555,c_2786
201760,0_1172,2021-10-05 22:29:18.194983,s_100555,c_2786
202508,0_1469,2021-10-06 12:32:24.528869,s_100849,c_5712
202513,0_1469,2021-10-06 12:37:05.198395,s_100849,c_5712
202972,0_999,2021-10-06 21:33:18.546110,s_101033,c_7755
202983,0_999,2021-10-06 21:55:57.913812,s_101033,c_7755


In [256]:
rows, cols = transactions.shape
nulls = transactions.isnull().sum()

summary_tx = pd.DataFrame({
    "Non-null": rows - nulls,
    "Valeurs manquantes": nulls,
    "% manquantes": (nulls/rows*100).round(2)
})
summary_tx

,Non-null,Valeurs manquantes,% manquantes
id_prod,687534,361041,34.43
date,687534,361041,34.43
session_id,687534,361041,34.43
client_id,687534,361041,34.43


<div style="
  font-family: Segoe UI, Arial;
  font-size: 16.5px;
  color: #eaeaea;
  background: linear-gradient(180deg, rgba(30,30,30,0.92), rgba(18,18,18,0.92));
  padding: 18px 20px;
  border-radius: 12px;
  border: 1px solid rgba(120,86,255,0.35);
  width: min(900px, 95%);
  margin: 22px auto;
  text-align: left;
  box-shadow: 0 0 18px rgba(0,0,0,0.45);
">

  <div style="display:flex; align-items:center; gap:10px; margin-bottom:8px;">
    <span style="font-size:18px;">🧾</span>
    <b style="font-size:18px;">
      Synthèse — Analyse exploratoire du fichier
      <span style="color:#ffcc00;">Transactions.csv</span>
    </b>
  </div>

  <hr style="border:none; border-top:1px solid rgba(255,255,255,0.08); margin:10px 0 14px 0;" />

  <p style="margin:8px 0;">
    📌 <b>Dimensions :</b>
    <span style="color:#ffcc00; font-weight:700;">1 048 575</span> lignes •
    <span style="color:#ffcc00; font-weight:700;">4</span> colonnes
    <span style="color:#9aa0a6;">(id_prod, date, session_id, client_id)</span>
  </p>

  <p style="margin:8px 0;">
    🧬 <b>Types :</b>
    <span style="color:#c9d1d9;">id_prod</span> (string),
    <span style="color:#c9d1d9;">date</span> (datetime),
    <span style="color:#c9d1d9;">session_id</span> (string),
    <span style="color:#c9d1d9;">client_id</span> (string)
  </p>

  <p style="margin:8px 0;">
    ⚠️ <b>Valeurs manquantes :</b>
    <span style="color:#ffcc00; font-weight:700;">361 041</span> valeurs manquantes par colonne
    <span style="color:#9aa0a6;">(soit</span>
    <span style="color:#ffcc00; font-weight:700;">34.43%</span>
    <span style="color:#9aa0a6;">des lignes)</span><br>
    <span style="color:#9aa0a6;">→ Ces lignes incomplètes devront être traitées avant les jointures.</span>
  </p>

  <p style="margin:8px 0;">
    🔗 <b>Cohérence (clés étrangères) — hors NaN :</b><br>
    • Transactions avec <span style="color:#ffcc00; font-weight:700;">id_prod</span> absent de <b>Products</b> :
    <span style="color:#34d399; font-weight:700;">0</span><br>
    • Transactions avec <span style="color:#ffcc00; font-weight:700;">client_id</span> absent de <b>Customers</b> :
    <span style="color:#34d399; font-weight:700;">0</span>
  </p>

  <p style="margin:8px 0;">
    🛒 <b>Achats multiples (cas métier) :</b><br>
    Analyse des répétitions d’un même produit acheté par un même client dans une même session :<br>
    • Nombre de cas avec répétition (<b>nb_occurrences &gt; 1</b>) :
      <span style="color:#ffcc00; font-weight:700;">853</span><br>
    • Maximum observé :
      <span style="color:#ffcc00; font-weight:700;">3</span><br>
    <span style="color:#9aa0a6;">→ Répétitions rares et plausibles : à conserver (ce n’est pas du bruit automatique).</span>
  </p>

  <p style="margin:8px 0;">
    🧹 <b>Doublons techniques :</b><br>
    • Doublons <b>exacts</b> (ligne complète identique) :
      <span style="color:#ffcc00; font-weight:700;">361 040</span><br>
    • Répétitions sur <span style="color:#ffcc00; font-weight:700;">(session_id, id_prod)</span> :
      <span style="color:#ffcc00; font-weight:700;">361 895</span><br>
    <span style="color:#9aa0a6;">→ Les doublons exacts sont de bons candidats à suppression ; les répétitions sur (session_id, id_prod) ne doivent pas être supprimées sans analyse (peut refléter des achats multiples).</span>
  </p>

  <div style="
    margin-top: 14px;
    padding: 10px 10px;
    border-radius: 10px;
    background-color: rgba(79,70,229,0.10);
    border: 1px solid rgba(79,70,229,0.25);
  ">
    ✅ <b>Conclusion :</b> le fichier <span style="color:#ffcc00;"><b>Transactions.csv</b></span> est
    <b>cohérent sur les références</b> (aucune clé étrangère manquante <i>hors NaN</i>).
    Avant les jointures, il faut <b>traiter les lignes incomplètes</b> (NaN) et <b>supprimer les doublons exacts</b>,
    tout en <b>conservant</b> les répétitions plausibles liées aux achats multiples.
  </div>

</div>

<a id="etape2_3_2"></a>
<div style="
  border: 1px solid #6D28D9;
  border-radius: 8px;
  background-color: rgba(109,40,217,0.06);
">
  <h3 style="
    margin: auto;
    padding: 12px 18px;
    color: #4F46E5;
    display: flex;
    align-items: center;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16" style="vertical-align:middle;
        filter: invert(28%) sepia(86%) saturate(2421%) hue-rotate(252deg) brightness(94%) contrast(96%);
      ">
    </a>
    2.3.2 - Traitements des doublons dans le fichier Transactions.csv
  </h3>
</div>

In [257]:
# ✅ État initial des transactions
# Objectif : mesurer le volume de données, identifier le nombre de lignes incomplètes
# (valeurs manquantes) et quantifier les doublons exacts (ligne complète identique)
# avant toute action de nettoyage, afin d'assurer la traçabilité

rows_before = transactions.shape[0]
nan_before = transactions.isnull().any(axis=1).sum()
dup_before = transactions.duplicated().sum()

rows_before, nan_before, dup_before

(1048575, np.int64(361041), np.int64(361040))

In [258]:
# 🧹 Nettoyage : suppression des lignes incomplètes
# Objectif : retirer les transactions contenant des valeurs manquantes sur les champs indispensables
# à la jointure (id_prod, date, session_id, client_id)
# ⚠️ Important : on ne supprime PAS ici des "doublons métier" (un client peut racheter un produit),
# seulement les lignes inutilisables pour l'analyse/jointure

transactions_clean = transactions.dropna(
    subset=["id_prod", "date", "session_id", "client_id"]
)

In [259]:
# 🔍 Contrôles post-nettoyage
# Objectif : vérifier que le dataset nettoyé ne contient plus de valeurs manquantes
# et qu'il ne reste aucun doublon exact, pour garantir une base saine avant jointure

transactions_clean.isnull().sum()
transactions_clean.duplicated().sum()

np.int64(0)

In [260]:
# ✅ État final après nettoyage
# Objectif : recalculer les indicateurs (lignes, NaN, doublons exacts) après nettoyage
# pour comparer "avant / après" et valider l'impact du traitement

rows_after = transactions_clean.shape[0]
nan_after = transactions_clean.isnull().any(axis=1).sum()
dup_after = transactions_clean.duplicated().sum()

rows_after, nan_after, dup_after

(687534, np.int64(0), np.int64(0))

In [261]:
# 📊 Synthèse avant / après
# Objectif : présenter de manière lisible l’impact du nettoyage :
# - nombre de lignes supprimées (lignes incomplètes),
# - disparition des valeurs manquantes,
# - évolution des doublons exacts

summary_cleaning = pd.DataFrame({
    "Avant nettoyage": [rows_before, nan_before, dup_before],
    "Après nettoyage": [rows_after, nan_after, dup_after]
}, index=[
    "Nombre de lignes",
    "Lignes avec valeurs manquantes",
    "Doublons exacts"
])

summary_cleaning

,Avant nettoyage,Après nettoyage
Nombre de lignes,1048575,687534
Lignes avec valeurs manquantes,361041,0
Doublons exacts,361040,0


<div style="
  font-family: Segoe UI, Arial;
  font-size: 16.5px;
  color: #eaeaea;
  background: linear-gradient(180deg, rgba(30,30,30,0.92), rgba(18,18,18,0.92));
  padding: 18px 20px;
  border-radius: 12px;
  border: 1px solid rgba(120,86,255,0.35);
  width: min(900px, 95%);
  margin: 22px auto;
  text-align: left;
  box-shadow: 0 0 18px rgba(0,0,0,0.45);
">

  <div style="display:flex; align-items:center; gap:10px; margin-bottom:8px;">
    <span style="font-size:18px;">🧹</span>
    <b style="font-size:18px;">
      Nettoyage — Transactions.csv
    </b>
  </div>

  <hr style="border:none; border-top:1px solid rgba(255,255,255,0.08); margin:10px 0 14px 0;" />

  <p style="margin:8px 0;">
    📌 <b>Objectif :</b> préparer la table <b>transactions</b> avant les jointures avec
    <b>customers</b> et <b>products</b> en supprimant les lignes inexploitables
    (valeurs manquantes sur les clés) et en contrôlant la présence de doublons exacts.
  </p>

  <div style="
    margin: 12px 0;
    padding: 12px 0px;
    border-radius: 10px;
    background-color: rgba(79,70,229,0.10);
    border: 1px solid rgba(79,70,229,0.25);
  ">
    <b>📊 Bilan chiffré (avant / après)</b><br><br>
    • <b>Nombre de lignes :</b>
    <span style="color:#ffcc00; font-weight:700;">1 048 575</span>
    → <span style="color:#34d399; font-weight:700;">687 534</span><br>
    • <b>Lignes avec valeurs manquantes (sur id_prod, date, session_id, client_id) :</b>
    <span style="color:#ffcc00; font-weight:700;">36 041</span>
    → <span style="color:#34d399; font-weight:700;">0</span><br>
    • <b>Doublons exacts (ligne complète identique) :</b>
    <span style="color:#ffcc00; font-weight:700;">361 040</span>
    → <span style="color:#34d399; font-weight:700;">0</span>
  </div>

  <p style="margin:8px 0;">
    ✅ <b>Interprétation :</b> la table nettoyée ne contient plus de valeurs manquantes sur les clés
    nécessaires à la jointure, et aucun doublon strict n’est présent dans les données conservées.
  </p>

  <div style="
    margin-top: 14px;
    padding: 10px 0px;
    border-radius: 10px;
    background-color: rgba(16,185,129,0.10);
    border: 1px solid rgba(16,185,129,0.25);
  ">
    ✅ <b>Conclusion :</b> <b>transactions_clean</b> est prête pour l’étape suivante : les jointures
    avec <b>customers</b> et <b>products</b>.
  </div>

</div>

<a id="etape3"></a>
<div style="
  border: 1px solid #F59E0B;
  border-radius: 8px;
  background-color: rgba(245, 158, 11, 0.12);
  margin: auto;
">
  <h2 style="
    margin: auto;
    padding: 15px 20px;
    background-color: #F59E0B;
    color: #fff;
    display: flex;
    align-items: center;
    border-radius: 8px;
    font-weight: 800;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="30" style="vertical-align:middle; filter: invert(8%) brightness(85%);">
    </a>
    Étape 3 — Construction du jeu de données d’analyse
  </h2>
</div>

<a id="etape3_1"></a>
<div style="border: 1px solid #F59E0B; border-radius:5px; background-color:rgba(245, 158, 11, 0.06);">
  <h3 style="margin:auto; padding:15px 20px; color:#FDBA74; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(77%) sepia(31%) saturate(1690%) hue-rotate(338deg) brightness(104%) contrast(98%);">
    </a>
    Étape 3.1 — Jointure transactions + produits
  </h3>
</div>


<a id="etape3_2"></a>
<div style="border: 1px solid #F59E0B; border-radius:5px; background-color:rgba(245, 158, 11, 0.06);">
  <h3 style="margin:auto; padding:15px 20px; color:#FDBA74; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(77%) sepia(31%) saturate(1690%) hue-rotate(338deg) brightness(104%) contrast(98%);">
    </a>
    Étape 3.2 — Jointure avec les données clients
  </h3>
</div>

<a id="etape3_3"></a>
<div style="border: 1px solid #F59E0B; border-radius:5px; background-color:rgba(245, 158, 11, 0.06);">
  <h3 style="margin:auto; padding:15px 20px; color:#FDBA74; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(77%) sepia(31%) saturate(1690%) hue-rotate(338deg) brightness(104%) contrast(98%);">
    </a>
    Étape 3.3 — Validation du dataset final
  </h3>
</div>

<a id="etape4"></a>
<div style="
  border: 1px solid #22C55E;
  border-radius: 8px;
  background-color: rgba(34, 197, 94, 0.10);
  margin: auto;
">
  <h2 style="
    margin: auto;
    padding: 15px 20px;
    background-color: #22C55E;
    color: #fff;
    display: flex;
    align-items: center;
    border-radius: 8px;
    font-weight: 800;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="30" style="vertical-align:middle; filter: invert(10%) brightness(90%);">
    </a>
    Étape 4 — Tops & flops produits
  </h2>
</div>

<a id="etape4_1"></a>
<div style="border: 1px solid #22C55E; border-radius:5px; background-color:rgba(34, 197, 94, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#86EFAC; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(84%) sepia(26%) saturate(846%) hue-rotate(84deg) brightness(103%) contrast(95%);">
    </a>
    Étape 4.1 — Évolution du chiffre d’affaires
  </h3>
</div>

<a id="etape4_2"></a>
<div style="border: 1px solid #22C55E; border-radius:5px; background-color:rgba(34, 197, 94, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#86EFAC; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(84%) sepia(26%) saturate(846%) hue-rotate(84deg) brightness(103%) contrast(95%);">
    </a>
    Étape 4.2 — Moyennes mobiles
  </h3>
</div>

<a id="etape4_3"></a>
<div style="border: 1px solid #22C55E; border-radius:5px; background-color:rgba(34, 197, 94, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#86EFAC; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(84%) sepia(26%) saturate(846%) hue-rotate(84deg) brightness(103%) contrast(95%);">
    </a>
    Étape 4.3 — Clients, transactions, produits par période
  </h3>
</div>

<a id="etape5"></a>
<div style="
  border: 1px solid #EC4899;
  border-radius: 8px;
  background-color: rgba(236, 72, 153, 0.10);
  margin: auto;
">
  <h2 style="
    margin: auto;
    padding: 15px 20px;
    background-color: #EC4899;
    color: #fff;
    display: flex;
    align-items: center;
    border-radius: 8px;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="30" style="vertical-align:middle; filter: invert(100%) brightness(115%);">
    </a>
    Étape 5 — Analyse de la performance des produits
  </h2>
</div>

<a id="etape5_1"></a>
<div style="border: 1px solid #EC4899; border-radius:5px; background-color:rgba(236, 72, 153, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#F9A8D4; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(74%) sepia(28%) saturate(1318%) hue-rotate(293deg) brightness(103%) contrast(98%);">
    </a>
    Étape 5.1 — Tops produits
  </h3>
</div>

<a id="etape5_2"></a>
<div style="border: 1px solid #EC4899; border-radius:5px; background-color:rgba(236, 72, 153, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#F9A8D4; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(74%) sepia(28%) saturate(1318%) hue-rotate(293deg) brightness(103%) contrast(98%);">
    </a>
    Étape 5.2 — Flops produits
</div>

<a id="etape5_3"></a>
<div style="border: 1px solid #EC4899; border-radius:5px; background-color:rgba(236, 72, 153, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#F9A8D4; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(74%) sepia(28%) saturate(1318%) hue-rotate(293deg) brightness(103%) contrast(98%);">
    </a>
    Étape 5.3 — Répartition du CA par catégorie
  </h3>
</div>

<a id="etape6"></a>
<div style="
  border: 1px solid #EAB308;
  border-radius: 8px;
  background-color: rgba(234, 179, 8, 0.12);
  margin: auto;
">
  <h2 style="
    margin: auto;
    padding: 15px 20px;
    background-color: #EAB308;
    color: #fff;
    display: flex;
    align-items: center;
    border-radius: 8px;
    font-weight: 900;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="30" style="vertical-align:middle; filter: invert(8%) brightness(85%);">
    </a>
    Étape 6 — Analyse des clients
  </h2>
</div>

<a id="etape6_1"></a>
<div style="border: 1px solid #EAB308; border-radius:5px; background-color:rgba(234, 179, 8, 0.06);">
  <h3 style="margin:auto; padding:15px 20px; color:#FDE68A; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(92%) sepia(21%) saturate(1023%) hue-rotate(333deg) brightness(104%) contrast(96%);">
    </a>
    Étape 6.1 — Profils clients & données démographiques
  </h3>
</div>

<a id="etape6_2"></a>
<div style="border: 1px solid #EAB308; border-radius:5px; background-color:rgba(234, 179, 8, 0.06);">
  <h3 style="margin:auto; padding:15px 20px; color:#FDE68A; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(92%) sepia(21%) saturate(1023%) hue-rotate(333deg) brightness(104%) contrast(96%);">
    </a>
    Étape 6.2 — Concentration du CA & courbe de Lorenz
  </h3>
</div>

<a id="etape6_2"></a>
<div style="border: 1px solid #EAB308; border-radius:5px; background-color:rgba(234, 179, 8, 0.06);">
  <h3 style="margin:auto; padding:15px 20px; color:#FDE68A; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(92%) sepia(21%) saturate(1023%) hue-rotate(333deg) brightness(104%) contrast(96%);">
    </a>
    Étape 6.3 — 
  </h3>
</div>

<a id="etape7"></a>
<div style="
  border: 1px solid #84CC16;
  border-radius: 8px;
  background-color: rgba(132, 204, 22, 0.10);
  margin: auto;
">
  <h2 style="
    margin: auto;
    padding: 15px 20px;
    background-color: #84CC16;
    color: #fff;
    display: flex;
    align-items: center;
    border-radius: 8px;
    font-weight: 900;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="30" style="vertical-align:middle; filter: invert(10%) brightness(90%);">
    </a>
    Étape 7 — Corrélations et tests statistiques
  </h2>
</div>

<a id="etape7_1"></a>
<div style="border: 1px solid #84CC16; border-radius:5px; background-color:rgba(132, 204, 22, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#BEF264; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(86%) sepia(41%) saturate(676%) hue-rotate(35deg) brightness(105%) contrast(96%);">
    </a>
    Étape 7.1 — Corrélations exploratoires
  </h3>
</div>

<a id="etape7_2"></a>
<div style="border: 1px solid #84CC16; border-radius:5px; background-color:rgba(132, 204, 22, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#BEF264; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(86%) sepia(41%) saturate(676%) hue-rotate(35deg) brightness(105%) contrast(96%);">
    </a>
    Étape 7.2 — Tests : χ², ANOVA (ou t-test), Pearson/Spearman
  </h3>
</div>

<a id="etape7_3"></a>
<div style="border: 1px solid #84CC16; border-radius:5px; background-color:rgba(132, 204, 22, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#BEF264; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(86%) sepia(41%) saturate(676%) hue-rotate(35deg) brightness(105%) contrast(96%);">
    </a>
    Étape 7.3 — Interprétation des résultats
  </h3>
</div>

<a id="etape8"></a>
<div style="
  border: 1px solid #EF4444;
  border-radius: 8px;
  background-color: rgba(239, 68, 68, 0.10);
  margin: auto;
">
  <h2 style="
    margin: auto;
    padding: 15px 20px;
    background-color: #EF4444;
    color: #fff;
    display: flex;
    align-items: center;
    border-radius: 8px;
    font-weight: 800;
  ">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="30" style="vertical-align:middle; filter: invert(100%) brightness(115%);">
    </a>
    Étape 8 — Clients atypiques (BtoB)
  </h2>
</div>

<a id="etape8_1"></a>
<div style="border: 1px solid #EF4444; border-radius:5px; background-color:rgba(239, 68, 68, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#FCA5A5; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(76%) sepia(23%) saturate(1182%) hue-rotate(314deg) brightness(103%) contrast(95%);">
    </a>
    Étape 8.1 — Identification des clients atypiques
  </h3>
</div>

<a id="etape8_2"></a>
<div style="border: 1px solid #EF4444; border-radius:5px; background-color:rgba(239, 68, 68, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#FCA5A5; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(76%) sepia(23%) saturate(1182%) hue-rotate(314deg) brightness(103%) contrast(95%);">
    </a>
    Étape 8.2 — Justification de l’exclusion pour l’analyse statistique
  </h3>
</div>

<a id="etape8_3"></a>
<div style="border: 1px solid #EF4444; border-radius:5px; background-color:rgba(239, 68, 68, 0.05);">
  <h3 style="margin:auto; padding:15px 20px; color:#FCA5A5; display:flex; align-items:center;">
    <a href="#top" style="text-decoration:none; display:flex; align-items:center; margin-right:10px;">
      <img src="assets/img/house-solid-full.svg" width="16"
           style="vertical-align:middle; filter: invert(76%) sepia(23%) saturate(1182%) hue-rotate(314deg) brightness(103%) contrast(95%);">
    </a>
    Étape 8.3 — Nouveau dataset “sans BtoB
  </h3>
</div>